### PR TITLE
Fix rrelu on CUDA

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #80434

Rrelu had a massive bug on CUDA, that would zero out the gradient in
the part where the function is the identity.

Fixes https://github.com/pytorch/pytorch/issues/80205

### CODE REVIEW

**Comment by Lezcano** [comment]
The main question here is: how did OpInfos failed to catch this?
I have inspected the OpInfo, and I have not been able to tell what's the reason why they didn't pick this up. There might be some incorrect handling of the `set_seed` wrapper that's being used in this function to handle the randomness.
Now, I have found two weird things I don't understand:
- Its `sample_kwargs` returns a `({'lower': 0., 'upper': 1.}, {'lower': 0., 'upper': 1.})`. I don't understand this, but this seems to be consistent with `generate_elementwise_unary_tensors`, which just takes the first element of `sample_kwargs`
- `gradcheck` for this operation just tests 3 inputs, none of them with kwargs...

**Comment by soulitzer** [comment]
> Its sample_kwargs returns a ({'lower': 0., 'upper': 1.}, {'lower': 0., 'upper': 1.})

This is used for the reference_inputs_func, one set of kwargs is used for numpy, and the other is for torch - this would matter if you need to pass in tensors for example. I wonder if it should also be used for sample_inputs_elementwise_unary by default, if op_kwargs is not specified @mruberry 

> gradcheck for this operation just tests 3 inputs, none of them with kwargs...

Yeah, if kwargs aren't passed, then `training=False` by default, and gradcheck does pass in that case. We probably want a sample_inputs_func for `rrelu` that passes in `training=True` as a kwarg.

**Comment by ngimel** [comment]
@Lezcano can you please add a test to OpInfo that would catch this. Note, for gradcheck to pass it will have to be an input of all positive values, otherwise with randomness results won't be deterministic

**Comment by soulitzer** [comment]
> Note, for gradcheck to pass it will have to be an input of all positive values, otherwise with randomness results won't be deterministic

That should be OK right? Don't we have a wrapper that freezes rng state and sets the seed?

**Comment by ngimel** [comment]
I don't know if freezing is enough for gradcheck (especially slow gradcheck) where forward would be called many times, each time advancing rng state

**Comment by soulitzer** [comment]
Isn't freezing rng state supposed to prevent it from being advanced? See below, both setting the seed and freezing the rng state seem to work:
```
from  torch.testing._internal.common_utils import freeze_rng_state
import torch

def fn(x):
    torch.manual_seed(42)
    return torch.rrelu(x, training=True)

def fn2(x):
    with freeze_rng_state():
        return torch.rrelu(x, training=True)

for device in ("cpu", "cuda"):  # Fails when CUDA
    a = torch.rand(10, 10, 10, requires_grad=True, dtype=torch.float64, device=device)

    for fast_mode in (True, False):  # Doesn't affect results
        print(fast_mode, device)
        print(torch.autograd.gradcheck(fn, (a,), fast_mode=fast_mode, raise_exception=False))
        print(torch.autograd.gradcheck(fn2, (a,), fast_mode=fast_mode, raise_exception=False))

```

**Comment by ngimel** [comment]
Ah, ok, even better

**Comment by Lezcano** [comment]
Done! 
A few notes on the process cc @mruberry 
Adding args / kwargs to ops in `UnaryUfuncInfo`s is truly dreadful. 
- I had to specify the kwargs in three different places to achieve this.
- In particular, if you do not know that specifying `sample_kwargs` does not specify them in the actual samles (????) then it turns out that you're not getting these kwargs in the tests (???!!!)
- It's not possible to add a set of kwargs. `ReductionOpInfo`s do a much better job here, taking an optional `generate_args_kwargs` for this.

Also:
- `wrapper_set_seed` is implemented as a forwarding operation that expects a function and its params. As such, you need to create an auxiliary lambda, and plenty of tests start complaining because you used a lambda. We could make it a decorator and that would save quite a bit of code, it'd make everything more readable (more functional) and perhaps these tests start passing.

File: aten/src/ATen/native/cuda/RreluWithNoise.cu [hunk 1]
```diff
@@ -155,7 +155,7 @@ Tensor& rrelu_with_noise_out_cuda(const Tensor& self,
   checkAllSameGPU("rrelu_with_noise_out_cuda", {self_arg, noise_arg, output_arg});
 
   if (training) {
-    AT_DISPATCH_FLOATING_TYPES_AND_HALF(
+    AT_DISPATCH_FLOATING_TYPES_AND2(at::ScalarType::Half, at::ScalarType::BFloat16,
```
Comment (Lezcano) [inline aten/src/ATen/native/cuda/RreluWithNoise.cu:158]: Support the same dtypes as `leaky_relu`.

**Comment by Lezcano** [comment]
I think this one should be ready. PTAL @ngimel @soulitzer

**Review by albanD — APPROVED** [review]
SGTM
